endpoints/data_objects/src/main.cpp

```diff
@@ -356,6 +356,100 @@ namespace
 		std::int64_t remaining_bytes_;
 	}; // incremental_read
 
+	class incremental_write : public std::enable_shared_from_this<incremental_write>
+	{
+	  public:
+		incremental_write(
+			irods::http::session_pointer_type& _sess_ptr,
+			unsigned int _http_version,
+			bool _http_keep_alive,
+			irods::http::connection_facade& _conn,
+			std::unique_ptr<io::client::native_transport>& _tp,
+			std::unique_ptr<io::odstream>& _out,
+			io::odstream* _out_ptr,
+			std::unique_ptr<irods::at_scope_exit<std::function<void()>>>& _mark_pw_stream_as_usable,
+			std::string& _buffer,
+			std::int64_t _remaining_bytes,
+			std::int64_t _internal_buffer_size,
+			bool _is_parallel_write)
+			: sess_ptr_{_sess_ptr->shared_from_this()}
+			, res_{http::status::ok, _http_version}
+			, conn_{std::move(_conn)}
+			, tp_{std::move(_tp)}
+			, out_{std::move(_out)}
+			, out_ptr_{_out_ptr}
+			, mark_pw_stream_as_usable_{std::move(_mark_pw_stream_as_usable)}
+			, buffer_{std::move(_buffer)}
+			, remaining_bytes_{_remaining_bytes}
+			, internal_buffer_size_{_internal_buffer_size}
+			, read_pos_{buffer_.data()}
+			, is_parallel_write_{_is_parallel_write}
+		{
+			res_.set(http::field::server, irods::http::version::server_name);
+			res_.set(http::field::content_type, "application/json");
+			res_.keep_alive(_http_keep_alive);
+		}
+
+		auto start() -> void
+		{
+			stream_bytes_to_irods();
+		} // start
+
+	  private:
+		auto stream_bytes_to_irods() -> void
+		{
+			irods::http::globals::background_task([self = shared_from_this(), fn = __func__]() mutable {
+				if (self->remaining_bytes_ > 0) {
+					if (!*self->out_ptr_) {
+						logging::error(
+							"{}: Output stream is in a bad state. Client should restart the entire transfer.", fn);
+						return self->sess_ptr_->send(
+							irods::http::fail(self->res_, http::status::internal_server_error));
+					}
+
+					const auto to_send = std::min<std::streamsize>(self->remaining_bytes_, self->internal_buffer_size_);
+					logging::debug(
+						"{}: Write buffer: remaining=[{}], sending=[{}].", fn, self->remaining_bytes_, to_send);
+					self->out_ptr_->write(self->read_pos_, to_send);
+					self->read_pos_ += to_send;
+					self->remaining_bytes_ -= to_send;
+
+					return self->stream_bytes_to_irods();
```

Comment:
So rather than having a single foreground task which loops to stream bytes, we've switched to a recursive background task which spawns background tasks from itself for each write (itself also being a background task) until there's nothing left to write. Is that right?

If so, why the switch to this model? Just trying to understand.

Reply:
The switch was inevitable. If we don't break up long running tasks, then the server runs the risk of becoming unresponsive. There are a limited number of threads and they have to devote time to each request. If too many clients appear and each of them attempt to upload a file which causes a thread to have to enter a loop, clients will feel the latency.

Reply:
Sure thing.

Reply:
Done.